File: frontend/src/api/telegram.ts
const API = import.meta.env.VITE_API_URL || "";

export const MY_TELEGRAM_APPS_URL = "https://my.telegram.org/apps";

export interface TelegramStatus {
  has_credentials: boolean;
  is_authorized: boolean;
  setup_complete: boolean;
  setup_step: "credentials" | "auth" | "complete";
  api_id: number | null;
  username: string | null;
  user_id: number | null;
  first_name: string | null;
  last_name: string | null;
  monitor_chat_id: number | null;
  app_title: string | null;
  qr_pending: boolean;
  my_telegram_apps_url: string;
}

export async function fetchTelegramStatus(): Promise<TelegramStatus> {
  const r = await fetch(`${API}/api/telegram/status`);
  if (!r.ok) throw new Error("Failed to load status");
  return r.json();
}

export async function saveTelegramCredentials(body: {
  api_id: number;
  api_hash: string;
  monitor_chat_id?: number;
  app_title?: string;
  app_short_name?: string;
}) {
  const r = await fetch(`${API}/api/telegram/credentials`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!r.ok) {
    const err = await r.json().catch(() => ({}));
    const detail = (err as { detail?: string | string[] }).detail;
    const msg = Array.isArray(detail) ? detail.join(", ") : detail;
    throw new Error(msg || `Ошибка сохранения (${r.status})`);
  }
  return r.json();
}

export async function startQrLogin() {
  const r = await fetch(`${API}/api/telegram/auth/qr/start`, { method: "POST" });
  if (!r.ok) {
    const err = await r.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail || "QR start failed");
  }
  return r.json() as Promise<{
    login_id: string;
    url: string | null;
    already_authorized: boolean;
    username?: string;
    expires_in?: number;
  }>;
}

export async function refreshQrLogin(loginId: string) {
  const r = await fetch(`${API}/api/telegram/auth/qr/refresh/${loginId}`, {
    method: "POST",
  });
  if (!r.ok) {
    const err = await r.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail || "QR refresh failed");
  }
  return r.json() as Promise<{
    login_id: string;
    url: string | null;
    expires_in?: number;
  }>;
}

export async function getQrStatus(loginId: string) {
  const r = await fetch(`${API}/api/telegram/auth/qr/status/${loginId}`);
  if (!r.ok) throw new Error("Status check failed");
  return r.json() as Promise<{
    status: string;
    message?: string;
    username?: string;
    user_id?: number;
    first_name?: string;
    last_name?: string;
  }>;
}

export function pollQrStatus(
  loginId: string,
  onUpdate: (s: { status: string; message?: string }) => void | Promise<void>
) {
  const id = setInterval(async () => {
    try {
      const s = await getQrStatus(loginId);
      await onUpdate(s);
      if (
        s.status === "authorized" ||
        s.status === "expired" ||
        s.status === "token_expired"
      ) {
        clearInterval(id);
      }
    } catch {
      /* retry */
    }
  }, 2000);
  return () => clearInterval(id);
}

export async function sendPhoneCode(phone: string) {
  const r = await fetch(`${API}/api/telegram/auth/phone/send`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ phone }),
  });
  if (!r.ok) {
    const err = await r.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail || "Send code failed");
  }
  return r.json() as Promise<{
    login_id: string;
    code_sent: boolean;
    phone_masked?: string;
    message?: string;
    already_authorized?: boolean;
  }>;
}

export async function verifyPhoneCode(
  loginId: string,
  code: string,
  password?: string
) {
  const r = await fetch(`${API}/api/telegram/auth/phone/verify`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ login_id: loginId, code, password: password || undefined }),
  });
  if (!r.ok) {
    const err = await r.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail || "Verify failed");
  }
  return r.json() as Promise<{
    status: string;
    message?: string;
    username?: string;
  }>;
}

export async function fetchTelegramCredentials() {
  const r = await fetch(`${API}/api/telegram/credentials`);
  if (r.status === 404 || r.status === 204) return null;
  if (!r.ok) throw new Error("Failed to load credentials");
  const text = await r.text();
  if (!text) return null;
  return JSON.parse(text) as {
    api_id: number;
    api_hash_masked: string;
    monitor_chat_id: number | null;
    app_title: string | null;
  };
}

export async function logoutTelegram() {
  const r = await fetch(`${API}/api/telegram/auth/logout`, { method: "POST" });
  if (!r.ok) throw new Error("Logout failed");
  return r.json() as Promise<{ ok: boolean; message?: string }>;
}

export async function resetTelegramAll() {
  const r = await fetch(`${API}/api/telegram/reset`, { method: "POST" });
  if (!r.ok) throw new Error("Reset failed");
  return r.json();
}

File: backend/app/schemas/telegram.py
from pydantic import BaseModel, Field


class TelegramCredentialsIn(BaseModel):
    api_id: int = Field(..., gt=0, description="App api_id from my.telegram.org/apps")
    api_hash: str = Field(..., min_length=16, description="App api_hash from my.telegram.org/apps")
    monitor_chat_id: int | None = None
    app_title: str | None = Field(None, description="App title from my.telegram.org (optional)")
    app_short_name: str | None = Field(None, description="Short name from my.telegram.org (optional)")


class TelegramCredentialsOut(BaseModel):
    api_id: int
    api_hash_masked: str
    monitor_chat_id: int | None
    app_title: str | None = None


class TelegramStatusOut(BaseModel):
    has_credentials: bool
    is_authorized: bool
    setup_complete: bool
    setup_step: str
    hosted_app: bool = False
    api_id: int | None
    username: str | None
    user_id: int | None
    first_name: str | None = None
    last_name: str | None = None
    monitor_chat_id: int | None
    app_title: str | None = None
    qr_pending: bool
    my_telegram_apps_url: str = "https://my.telegram.org/apps"


class QrStartOut(BaseModel):
    login_id: str
    url: str | None
    already_authorized: bool = False
    username: str | None = None
    expires_in: int | None = None
    refreshed: bool = False


class QrStatusOut(BaseModel):
    status: str
    login_id: str | None = None
    username: str | None = None
    user_id: int | None = None
    first_name: str | None = None
    last_name: str | None = None
    message: str | None = None


class PhoneSendIn(BaseModel):
    phone: str = Field(..., description="Номер с кодом страны, например +79991234567")


class PhoneSendOut(BaseModel):
    login_id: str
    code_sent: bool
    phone_masked: str | None = None
    message: str | None = None
    code_delivery: str | None = Field(
        None,
        description="app | sms | call | email — куда Telegram отправил код",
    )
    delivery_hint: str | None = None
    already_authorized: bool = False
    username: str | None = None


class PhoneVerifyIn(BaseModel):
    login_id: str
    code: str = Field(..., min_length=4, max_length=10)
    password: str | None = Field(None, description="Облачный пароль 2FA, если включён")


class PhoneVerifyOut(BaseModel):
    status: str
    login_id: str | None = None
    username: str | None = None
    user_id: int | None = None
    first_name: str | None = None
    last_name: str | None = None
    message: str | None = None
